src/xbot/common/subsystems/BaseDriveSubsystem.java

```diff
@@ -0,0 +1,142 @@
+package xbot.common.subsystems;
+
+import java.util.Map;
+import com.ctre.CANTalon.TalonControlMode;
+
+import xbot.common.command.BaseSubsystem;
+import xbot.common.controls.actuators.XCANTalon;
+import xbot.common.logging.LoggingLatch;
+import xbot.common.logic.Latch.EdgeType;
+import xbot.common.math.XYPair;
+
+public abstract class BaseDriveSubsystem extends BaseSubsystem {
+    
+    private final LoggingLatch baseDriveSubsystemLoggingLatch = 
+            new LoggingLatch(this.getName(), "XCanTalon(s) in DriveSubsystem is null", EdgeType.RisingEdge);
+    
+    public class MotionRegistration {
+
+        public double x;
+        public double y;
+        public double w;
+        
+        public MotionRegistration(double x, double y, double w) {
+            this.x = x;
+            this.y = y;
+            this.w = w;
+        }        
+        
+        public double calculateTotalImpact(double x, double y, double w) {
+            return this.x * x + this.y*y + this.w * w;
+        }
+    }
+    
+    /**
+     * This is the "heart" of the DriveSubsystem - the rest of the class depends on returning this
+     * list of Talons and how they respond to drive inputs.
+     */
+    protected abstract Map<XCANTalon, MotionRegistration> getAllMasterTalons();
+    
+    /**
+     * Returns the total distance tracked by the encoder
+     * - On the LEFT side of the robot
+     * - Pointing in the direction of +Y travel
+     */
+    public abstract double getLeftTotalDistance();
```

Comment:
How do you envision these three operating for tank and mecanum? Is transverse always zero on a tank robot?

Reply:
The DriveSubsystem is in charge of whatever encoders are related to drive - sometimes that's gearbox encoders, and sometimes we have "follow" encoders.

If we only have gearbox encoders, transverse would always be 0 for a tank robot.

However, if we had follow-wheel encoders, we could install a transverse encoder, which would be able to detect our robot being pushed or slipping, and so we would have some transverse value.